File: ruby/custom-set/custom_set.rb
class CustomSet
  def initialize(array)
    @buffer = array
  end

  def size
    @buffer.size
  end

  def empty?
    @buffer.empty?
  end

  def each
    @buffer.each { |i| yield i }
  end

  def member?(item)
    @buffer.include?(item)
  end

  def subset?(other)
    @buffer.all? { |i| other.member?(i) }
  end

  def disjoint?(other)
    @buffer.none? { |i| other.member?(i) }
  end

  def ==(other)
    subset?(other) && size == other.size
  end

  def add(item)
    @buffer << item unless member?(item)
    return self
  end

  def intersection(other)
    CustomSet.new(
      @buffer.select { |i| other.member?(i) }
    )
  end

  def difference(other)
    CustomSet.new(
      @buffer.reject { |i| other.member?(i) }
    )
  end

  def union(other)
    out = CustomSet.new(@buffer)
    other.each { |i| out.add(i) }
    return out
  end
end

File: javascript/custom-set/custom-set.js
export class CustomSet {
    constructor(array) {
        if (array)
            this.buffer = [...array];
        else
            this.buffer = [];
    }

    get size() {
        return this.buffer.length;
    }

    empty() {
        return this.size == 0;
    }

    contains(n) {
        return this.buffer.includes(n);
    }

    add(n) {
        if (!this.contains(n))
            this.buffer.push(n);

        return this;
    }

    subset(other) {
        return this.buffer.every(x => other.contains(x));
    }

    disjoint(other) {
        return this.buffer.every(x => !other.contains(x));
    }

    eql(other) {
        return this.subset(other) && this.size == other.size;
    }

    union(other) {
        let output = new CustomSet(this.buffer);

        other.buffer.forEach(y => output.add(y));

        return output;
    }

    intersection(other) {
        let output = new CustomSet;

        for (const x of this.buffer)
            if (other.contains(x))
                output.add(x);

        return output;
    }

    difference(other) {
        let output = new CustomSet;

        for (const x of this.buffer)
            if (!other.contains(x))
                output.add(x);

        return output;
    }
}
